bookstore/client/notebook.py

```diff
@@ -0,0 +1,165 @@
+    #     last_activity: str #'2019-03-14T23:38:08.137987Z',
+    #     execution_state: str #'idle',
+    #     connections: int # 0
+    def __init__(self, *args, id, name, last_activity, execution_state, connections):
+        self.id = id
+        self.name = name
+        self.last_activity = last_activity
+        self.execution_state = execution_state
+        self.connections = connections
+
+
+class NotebookSession:  # (NamedTuple):
+    #     id: str #'68d9c58f-c57d-4133-8b41-5ec2731b268d',
+    #     path: str #'Untitled38.ipynb',
+    #     name: str #'',
+    #     type: str #'notebook',
+    #     kernel: KernelInfo
+    #     notebook: dict # {'path': 'Untitled38.ipynb', 'name': ''}}}
+
+    def __init__(self, *args, path, name, type, kernel, notebook, **kwargs):
+        self.path = path
+        self.name = name
+        self.type = type
+        self.kernel = KernelInfo(**kernel)
+        self.notebook = notebook
+
+
+class NotebookClient:
+    def __init__(self, nb_config):
+        self.nb_config = nb_config
+        self.nb_record = LiveNotebookRecord(**self.nb_config)
+        self.url = self.nb_record.url.rstrip(
+            "/"
+        )  # So that we can have full API endpoints without double //
+        self.token = self.nb_record.token
+        sessions_temp = self.get_sessions()
+        self.sessions = {session['kernel']['id']: session for session in sessions_temp}
+
+    @property
+    def sessions_endpoint(self):
+        api_endpoint = "/api/sessions/"
+        return f"{self.url}{api_endpoint}"
+
+    def get_sessions(self):
+        target_url = f"{self.sessions_endpoint}"
+        headers = {'Authorization': f'token {self.token}'}
+        resp = requests.get(target_url, headers=headers)
+        return resp.json()
+
+    @property
+    def kernels_endpoint(self):
+        api_endpoint = "/api/kernels/"
+        return f"{self.url}{api_endpoint}"
+
+    def get_kernels(self):
+        target_url = f"{self.sessions_endpoint}"
+        headers = {'Authorization': f'token {self.token}'}
+        resp = requests.get(target_url, headers=headers)
+        return resp.json()
+
+    @property
+    def contents_endpoint(self):
+        api_endpoint = "/api/contents/"
+        return f"{self.url}{api_endpoint}"
+
+    def get_contents(self, path):
+        target_url = f"{self.contents_endpoint}{path}"
+        headers = {'Authorization': f'token {self.token}'}
+        resp = requests.get(target_url, headers=headers)
+        return resp.json()
+
+
+def python_compat_session(session):
+    deepcopy(session)
+
+
+class NotebookClientCollection:
+    nb_client_gen = lambda: (NotebookClient(x) for x in list_running_servers())
+    sessions = {x.url: x.sessions for x in nb_client_gen()}
+
+    @classmethod
+    def current_server(cls):
+        for server_url, session_dict in cls.sessions.items():
+            for session_id, session in session_dict.items():
+                python_compat_session(session)
```

Comment:
oops — I had actually intended to make a copy of the session there, though now in retrospect I don't need to so I can clean that up.